### ISSUE TITLE
Proposal: add `#![deny(unused_crate_dependencies)]` to all single target package

### ISSUE THREAD

**Issue by pymongo** [datafuselabs/databend#7016]
**Summary**

cargo-udeps is not precise to analysis unused dependency

e.g. when I use this lint: <https://doc.rust-lang.org/rustc/lints/listing/allowed-by-default.html#unused-crate-dependencies>

to analysis this crate common/meta/app

**cargo-udeps said all deps is used, but rustc said 15 deps is unused just increase compile time**

```
[w@ww databend]$ cargo udeps
info: Loading save analysis from "/home/w/repos/database/databend/target/debug/deps/save-analysis/databend_query-22cb56d1d8dc7d68.json"
info: Ignoring `openssl-src` (Normal)
info: Ignoring `axum` (Normal)
All deps seem to have been used.
[w@ww databend]$ cargo c
    Checking common-meta-app v0.1.0 (/home/w/repos/database/databend/common/meta/app)
error: external crate `anyerror` unused in `common_meta_app`: remove the dependency or add `use anyerror as _;`
  --> common/meta/app/src/lib.rs:15:1
   |
15 | //! This crate defines meta data types used by meta-client application, e.g. Schema, User, Share etc.
   | ^
   |
note: the lint level is defined here
  --> common/meta/app/src/lib.rs:21:9
   |
21 | #![deny(unused_crate_dependencies)]
...
error: external crate `thiserror` unused in `common_meta_app`: remove the dependency or add `use thiserror as _;`
error: external crate `tonic` unused in `common_meta_app`: remove the dependency or add `use tonic as _;`
error: could not compile `common-meta-app` due to 15 previous errors
```

When I search text `thiserror` in common/meta/app, there is no results
thiserror is unused in common-meta-app but cargo-udeps can't report this unused,
we can use `#![deny(unused_crate_dependencies)]` to found this unused

## why only enable on single target package

`#![deny(unused_crate_dependencies)]` work well is single target crate,

If one crate/package has multi target
e.g. main.rs use dep1 only and lib.rs use dep2 only
unused_crate_dependencies would report a error(this is a false positive)

Here is single target package list in databend workspace:

```
common-arrow
common-building
common-catalog
common-config
common-tracing
common-contexts
common-fuse-meta
common-meta-app
common-settings
common-meta-api
common-meta-store
common-macros
common-storages-hive
common-pipeline-sources
common-storages-index
common-pipeline-transforms
common-storages-util
common-legacy-parser
common-pipeline-sinks
common-storages-fuse
databend-metabench
databend-metactl
common-dep-hack
```

Here is my code to find single target package

```rust
let config = cargo::Config::default().unwrap();
let path = std::path::Path::new("/home/w/repos/database/databend/Cargo.toml");
let path = path.canonicalize().unwrap();
let workspace = cargo::core::Workspace::new(path.as_path(), &config).unwrap();
for package in workspace.members() {
    if package.targets().len() == 1 {
        println!("{}", package.name());
    }
}
```

In conclusion, the way to analysics unused deps:
- single target pakcage use `#![deny(unused_crate_dependencies)]`
- multi target package use `cargo-udeps`

**Comment by Xuanwo** [comment]
LGTM, are you willing to implement this proposal?

Also cc @PsiACE

**Comment by PsiACE** [comment]
I think this is a good proposal, although these unused dependencies may still exist in the dependency tree.
